apps/prairielearn/src/models/course-permissions.ts

```diff
@@ -0,0 +1,192 @@
+    { course_id, user_id, course_role, authn_user_id },
+    CoursePermissionSchema,
+  );
+  if (!result) {
+    throw new Error('No course permissions to update');
+  }
+}
+
+export async function deleteCoursePermissions({
+  course_id,
+  user_id,
+  authn_user_id,
+}: {
+  course_id: string;
+  user_id: string;
+  authn_user_id: string;
+}): Promise<void> {
+  await queryAsync(sql.delete_course_permissions, { course_id, user_id, authn_user_id });
+  // Do not throw an exception if no course permissions to delete
+}
+
+export async function deleteCoursePermissionsForNonOwners({
+  course_id,
+  authn_user_id,
+}: {
+  course_id: string;
+  authn_user_id: string;
+}): Promise<void> {
+  await runInTransactionAsync(async () => {
+    const nonOwners = await queryRows(
+      sql.select_and_lock_non_owners,
+      { course_id },
+      CoursePermissionSchema,
+    );
+    for (const nonOwner of nonOwners) {
+      await deleteCoursePermissions({ course_id, user_id: nonOwner.user_id, authn_user_id });
+    }
+  });
+}
+
+export async function deleteCoursePermissionsForUsersWithoutAccess({
+  course_id,
+  authn_user_id,
+}: {
+  course_id: string;
+  authn_user_id: string;
+}): Promise<void> {
+  await runInTransactionAsync(async () => {
+    const usersWithoutAccess = await queryRows(
+      sql.select_and_lock_course_permissions_without_access,
+      { course_id },
+      CoursePermissionSchema,
+    );
+    for (const user of usersWithoutAccess) {
+      await deleteCoursePermissions({ course_id, user_id: user.user_id, authn_user_id });
+    }
+  });
+}
+
+export async function insertCourseInstancePermissions({
+  course_id,
+  course_instance_id,
+  user_id,
+  course_instance_role,
+  authn_user_id,
+}: {
+  course_id: string;
+  course_instance_id: string;
+  user_id: string;
+  course_instance_role: NonNullable<CourseInstancePermission['course_instance_role']>;
+  authn_user_id: string;
+}): Promise<void> {
+  const coursePermission = await queryOptionalRow(
+    sql.insert_course_instance_permissions,
+    { course_id, course_instance_id, user_id, course_instance_role, authn_user_id },
+    CoursePermissionSchema,
+  );
+  if (!coursePermission) {
+    throw new Error('Cannot add permissions for a course instance without course permissions');
+  }
+}
+
+export async function updateCourseInstancePermissionsRole({
+  course_id,
+  course_instance_id,
+  user_id,
+  course_instance_role,
+  authn_user_id,
+}: {
+  course_id: string;
+  course_instance_id: string;
+  user_id: string;
+  course_instance_role: NonNullable<CourseInstancePermission['course_instance_role']>;
+  authn_user_id: string;
+}): Promise<void> {
+  const result = await queryOptionalRow(
+    sql.update_course_instance_permissions_role,
+    { course_id, course_instance_id, user_id, course_instance_role, authn_user_id },
+    CourseInstancePermissionSchema,
+  );
+  if (!result) {
+    throw new Error('No course instance permissions to update');
+  }
```

Comment:
I was a bit torn about what kind of error should be thrown here, since it's not expected to happen in normal circumstances, but 404 makes sense. Done in aa06dfc243df8480dec61e1a4cbb85a98f986a34 for all errors in this file.